Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,160 @@
+  `x * (sSup s) = ⨆ y ∈ s, x * y` and `(sSup s) * y = ⨆ x ∈ s, x * y`;
+
+* mix-in class `IsIntegral`: a unital quantale (i.e. a quantale in which the semigroup is
+  a monoid) is called integral, or strictly two-sided, when `⊤ = 1`;
+
+* next to these classes, we define the additive versions `AddQuantale`, `IsAddIntegral` in
+  which the semigroup is denoted `+` i.s.o. `*`;
+
+## Naming conventions
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup { z | z * x ≤ y }`, the left-residuation of `y` over `x`;
+
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the right-residuation of `y` over `x`;
+
+## References
+
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale>
+
+## TODO
+
+-/
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class AddQuantale (α : Type*) [AddSemigroup α] extends CompleteLattice α where
+  /-- Addition is distributive over join in a quantale -/
+  protected add_sSup_eq_iSup_add (x : α) (s : Set α) : x + sSup s = ⨆ y ∈ s, x + y
+  /-- Addition is distributive over join in a quantale -/
+  protected sSup_add_eq_iSup_add (s : Set α) (y : α) : sSup s + y = ⨆ x ∈ s, x + y
+
+/-- A quantale is a semigroup distributing over a complete lattice. -/
+@[to_additive]
+class Quantale (α : Type*) [Semigroup α] extends CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+/-- An integral (or strictly two-sided) additive quantale is a quantale over an additive monoid
+    where top and unit coincide. -/
+class IsIntegralAddQuantale (α : Type*) [AddMonoid α] [AddQuantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_zero : (⊤ : α) = 0
+
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+    top and unit coincide. -/
+@[to_additive]
+class IsIntegralQuantale (α : Type*) [Monoid α] [Quantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_one : (⊤ : α) = 1
+
+section Quantale
+
+variable (α : Type _)
+variable [Semigroup α] [Quantale α]
+
+@[to_additive]
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+
+@[to_additive]
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
+
+end Quantale
+
+section IsIntegral
+open Quantale
+
+variable (α : Type _)
+variable [Monoid α] [Quantale α] [IsIntegralQuantale α]
+
+@[to_additive]
+theorem top_eq_one : (⊤ : α) = 1 := IsIntegralQuantale.top_eq_one
+
+end IsIntegral
+
+namespace Quantale
+
+variable {α : Type _}
```

Comment:
```suggestion
variable {α : Type*}
```
and everywhere else

Reply:
Happy to comply, but struggling to understand the difference.
Can you explain what is what?